File: packages/sdk/src/utils/http.ts
import type { HttpMethod, HttpResponse, OneSubConfig } from '../types.js';
import { NetworkError, TimeoutError, parseApiError } from './errors.js';

/**
 * HTTP client for 1Sub API with retry logic and error handling
 */
export class HttpClient {
  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly timeout: number;
  private readonly maxRetries: number;
  private readonly debug: boolean;

  constructor(config: OneSubConfig) {
    this.apiKey = config.apiKey;
    this.baseUrl = config.baseUrl || 'https://1sub.io/api/v1';
    this.timeout = config.timeout || 30000;
    this.maxRetries = config.maxRetries ?? 3;
    this.debug = config.debug || false;
  }

  /**
   * Make an HTTP request with automatic retries
   */
  async request<T>(
    method: HttpMethod,
    path: string,
    body?: unknown
  ): Promise<HttpResponse<T>> {
    const url = `${this.baseUrl}${path}`;
    let lastError: Error | undefined;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      try {
        const response = await this.makeRequest<T>(method, url, body);
        return response;
      } catch (error) {
        lastError = error as Error;

        // Don't retry on client errors (4xx) except rate limits
        if (error instanceof Error && 'statusCode' in error) {
          const statusCode = (error as { statusCode: number }).statusCode;
          if (statusCode >= 400 && statusCode < 500 && statusCode !== 429) {
            throw error;
          }
        }

        // Exponential backoff for retries
        if (attempt < this.maxRetries) {
          const delay = this.calculateBackoff(attempt);
          this.log(`Request failed, retrying in ${delay}ms (attempt ${attempt + 1}/${this.maxRetries})`);
          await this.sleep(delay);
        }
      }
    }

    throw lastError || new NetworkError('Request failed after retries');
  }

  /**
   * Make a single HTTP request
   */
  private async makeRequest<T>(
    method: HttpMethod,
    url: string,
    body?: unknown
  ): Promise<HttpResponse<T>> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      this.log(`${method} ${url}`);

      const headers: Record<string, string> = {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json',
        'User-Agent': '@1sub/sdk/1.0.0',
      };

      const fetchOptions: RequestInit = {
        method,
        headers,
        signal: controller.signal,
      };

      if (body && (method === 'POST' || method === 'PUT' || method === 'PATCH')) {
        fetchOptions.body = JSON.stringify(body);
      }

      const response = await fetch(url, fetchOptions);

      // Parse response headers
      const responseHeaders: Record<string, string> = {};
      response.headers.forEach((value, key) => {
        responseHeaders[key.toLowerCase()] = value;
      });

      // Parse response body
      let data: T;
      const contentType = response.headers.get('content-type');
      if (contentType?.includes('application/json')) {
        data = await response.json() as T;
      } else {
        data = await response.text() as T;
      }

      // Handle error responses
      if (!response.ok) {
        parseApiError(response.status, data);
      }

      this.log(`Response: ${response.status}`);

      return {
        data,
        status: response.status,
        headers: responseHeaders,
      };
    } catch (error) {
      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          throw new TimeoutError(this.timeout);
        }
        if (error.message.includes('fetch')) {
          throw new NetworkError(error.message);
        }
      }
      throw error;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  /**
   * Calculate exponential backoff delay
   */
  private calculateBackoff(attempt: number): number {
    // Base delay of 1 second with exponential increase and jitter
    const baseDelay = 1000;
    const maxDelay = 30000;
    const exponentialDelay = baseDelay * Math.pow(2, attempt);
    const jitter = Math.random() * 1000;
    return Math.min(exponentialDelay + jitter, maxDelay);
  }

  /**
   * Sleep for specified milliseconds
   */
  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Log message if debug mode is enabled
   */
  private log(message: string): void {
    if (this.debug) {
      console.log(`[1Sub SDK] ${message}`);
    }
  }

  /**
   * GET request helper
   */
  async get<T>(path: string): Promise<HttpResponse<T>> {
    return this.request<T>('GET', path);
  }

  /**
   * POST request helper
   */
  async post<T>(path: string, body?: unknown): Promise<HttpResponse<T>> {
    return this.request<T>('POST', path, body);
  }

  /**
   * PUT request helper
   */
  async put<T>(path: string, body?: unknown): Promise<HttpResponse<T>> {
    return this.request<T>('PUT', path, body);
  }

  /**
   * DELETE request helper
   */
  async delete<T>(path: string): Promise<HttpResponse<T>> {
    return this.request<T>('DELETE', path);
  }
}

File: packages/python-sdk/onesub/http.py
"""1Sub SDK HTTP Client"""

import time
import random
from typing import Any, Dict, Optional

import requests

from .errors import parse_api_error, OneSubError


class HttpClient:
    """HTTP client for 1Sub API with retry logic"""

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://1sub.io/api/v1",
        timeout: int = 30,
        max_retries: int = 3,
        debug: bool = False
    ):
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.max_retries = max_retries
        self.debug = debug
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
            "User-Agent": "onesub-python/1.0.0",
        })

    def _log(self, message: str) -> None:
        if self.debug:
            print(f"[1Sub SDK] {message}")

    def _calculate_backoff(self, attempt: int) -> float:
        """Calculate exponential backoff delay"""
        base_delay = 1.0
        max_delay = 30.0
        delay = base_delay * (2 ** attempt)
        jitter = random.random()
        return min(delay + jitter, max_delay)

    def request(
        self,
        method: str,
        path: str,
        data: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Make an HTTP request with automatic retries"""
        url = f"{self.base_url}{path}"
        last_error: Optional[Exception] = None

        for attempt in range(self.max_retries + 1):
            try:
                self._log(f"{method} {url}")

                response = self.session.request(
                    method=method,
                    url=url,
                    json=data,
                    timeout=self.timeout
                )

                self._log(f"Response: {response.status_code}")

                if response.ok:
                    return response.json()

                # Parse error response
                try:
                    body = response.json()
                except ValueError:
                    body = {"message": response.text}

                # Don't retry client errors (except rate limits)
                if 400 <= response.status_code < 500 and response.status_code != 429:
                    parse_api_error(response.status_code, body)

                parse_api_error(response.status_code, body)

            except OneSubError:
                raise
            except requests.exceptions.Timeout:
                last_error = OneSubError(
                    f"Request timed out after {self.timeout}s",
                    "TIMEOUT",
                    0
                )
            except requests.exceptions.RequestException as e:
                last_error = OneSubError(str(e), "NETWORK_ERROR", 0)

            # Retry with backoff
            if attempt < self.max_retries:
                delay = self._calculate_backoff(attempt)
                self._log(f"Retrying in {delay:.1f}s (attempt {attempt + 1}/{self.max_retries})")
                time.sleep(delay)

        if last_error:
            raise last_error
        raise OneSubError("Request failed after retries", "NETWORK_ERROR", 0)

    def get(self, path: str) -> Dict[str, Any]:
        return self.request("GET", path)

    def post(self, path: str, data: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        return self.request("POST", path, data)

    def close(self) -> None:
        """Close the HTTP session"""
        self.session.close()
